cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/CoreSchedulerService.java

```diff
@@ -0,0 +1,189 @@
+      public Void run(ProgramScheduleStoreDataset store) throws AlreadyExistsException {
+        store.addSchedules(schedules);
+        return null;
+      }
+    }, AlreadyExistsException.class);
+  }
+
+  @Override
+  public void deleteSchedule(ScheduleId scheduleId) {
+    deleteSchedules(Collections.singleton(scheduleId));
+  }
+
+  @Override
+  public void deleteSchedules(final Iterable<? extends ScheduleId> scheduleIds) {
+    execute(new StoreTxRunnable<Void, RuntimeException>() {
+      @Override
+      public Void run(ProgramScheduleStoreDataset store) {
+        store.deleteSchedules(scheduleIds);
+        return null;
+      }
+    }, null);
+  }
+
+  @Override
+  public void deleteSchedules(final ApplicationId appId) {
+    execute(new StoreTxRunnable<Void, RuntimeException>() {
+      @Override
+      public Void run(ProgramScheduleStoreDataset store) {
+        store.deleteSchedules(appId);
+        return null;
+      }
+    }, null);
+  }
+
+  @Override
+  public ProgramSchedule getSchedule(final ScheduleId scheduleId) throws NotFoundException {
+    return execute(new StoreTxRunnable<ProgramSchedule, NotFoundException>() {
+      @Override
+      public ProgramSchedule run(ProgramScheduleStoreDataset store) throws NotFoundException {
+        return store.getSchedule(scheduleId);
+      }
+    }, NotFoundException.class);
+  }
+
+  @Override
+  public List<ProgramSchedule> listSchedules(final ApplicationId appId) {
+    return execute(new StoreTxRunnable<List<ProgramSchedule>, RuntimeException>() {
+      @Override
+      public List<ProgramSchedule> run(ProgramScheduleStoreDataset store) {
+        return store.listSchedules(appId);
+      }
+    }, null);
+  }
+
+  @Override
+  public List<ProgramSchedule> listSchedules(final ProgramId programId) {
+    return execute(new StoreTxRunnable<List<ProgramSchedule>, RuntimeException>() {
+      @Override
+      public List<ProgramSchedule> run(ProgramScheduleStoreDataset store) {
+        return store.listSchedules(programId);
+      }
+    }, null);
+  }
+
+  private interface StoreTxRunnable<V, T extends Throwable> {
+    V run(ProgramScheduleStoreDataset store) throws T;
+  }
+
+  private <V, T extends Exception> V execute(final StoreTxRunnable<V, T> runnable,
+                                             final Class<? extends T> tClass) throws T {
+    final AtomicReference<V> v = new AtomicReference<>();
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          ProgramScheduleStoreDataset store = context.getDataset(STORE_DATASET_ID.getDataset());
+          v.set(runnable.run(store));
+        }
+      });
+    } catch (TransactionConflictException | TransactionNotInProgressException e) {
+      throw Throwables.propagate(e);
+    } catch (TransactionFailureException e) {
+      if (e.getCause() == null || !e.getClass().equals(TransactionFailureException.class)) {
```

Comment:
You can use one of the `Transactions.propagate` method

https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java#L348